src/dbcp/transform/acp_projects.py

```diff
@@ -479,4 +486,64 @@ def _str_strip(ser):
     }
     out.rename(columns=rename_dict, inplace=True)
     out.drop(columns=cols_to_drop, inplace=True)
-    return {"acp_projects": out}
+    return out
+
+
+def _clean_col_names_and_create_id(df) -> pd.DataFrame:
+    """Create surrogate key based on raw columns, then rename and assign ID.
+
+    The surrogate key is created based on raw data and column names.
+    Create the surrogate key, then clean column names and specify
+    which columns are raw, and assign the surrogate key ID column.
+    """
+    surrogate_key = _make_surrogate_key(df)  # uses raw column names
+    df.columns = _rename_columns(df.columns)
+    df["proj_id"] = surrogate_key  # assign after renaming columns
+    return df
+
+
+def _transform_acp_projects(raw_df: pd.DataFrame) -> pd.DataFrame:
+    trans = raw_df.convert_dtypes()
+    trans = _clean_col_names_and_create_id(trans)
+    out = _clean_columns(trans)
+
+    return out
+
+
+def _transform_acp_snapshots_to_changelog(
+    raw_snapshots_df: pd.DataFrame,
+) -> pd.DataFrame:
+    """Create a changelog of the ACP data over time."""
+    trans_df = raw_snapshots_df.reset_index(drop=True).convert_dtypes()
+    trans_df = (
+        trans_df.groupby("report_date")
+        .apply(_clean_col_names_and_create_id)
```

Comment:
this isn't fast, but it's also not that slow. i like your suggestion to use a SQL query + window functions instead but I don't really trust my SQL knowledge enough to attempt to refactor `_clean_col_names_and_create_id` / thought I'd just be wasting time trying to figure it out. thoughts on how hard this might be?